### PR TITLE
Single source of truth for guard logging

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107567
* #107439
* #107471
* #107562
* __->__ #107532
* #107530
* #107516
* #107505

Instead of (poorly) reconstructing the guard list from the guards on OutputGraph, we log them at the horses mouth: when we actually codegen the guard. This only requires very modest refactoring: as we translate guards into code parts, we also have to pass the source guard along so we can use it to give stack information.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Comment by mlazos** [comment]
Why is this better than the status quo? Won't this yield the same result as iterating through code parts?

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -193,6 +216,7 @@ def __init__(
         # check it against this example", and it all ends up getting
         # swept up into a single call to ___check_tensors.  Invariant:
         # len(tensor_check_names) == len(tensor_check_examples).
+        # TODO: something here
```
Comment (anijain2305) [inline torch/_dynamo/guards.py:219]: Missing?

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/guards.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/guards.py:219]: Later in the stack

**Pull request closed without merging**